asllib/doc/ASLTypingReference.tex

```diff
@@ -495,149 +681,276 @@ \section{Domain of a Type}
+We say that $\dynamicdomain(\env, \vt)$ is the \emph{dynamic domain} of $\vt$
+in the environment $\env$.
 %
 The \emph{static domain} of a type is the set of values which storage elements of that type may hold
 \underline{across all possible dynamic environments}.
 %
 The reason that for this distinction is that the sets of values
 considered for bitvector types, array types, constrained integers, and constrained bitvectors,
-can be parameterized by configuration variables and subprogram parameters whose values are determined dynamically.
+can depend on the dynamic values of variables.
 
-Types that do not refer to subprogram parameters have a static domain that is equal to their dynamic domain.
+Types that do not refer to variables whose values are only known dynamically have
+a static domain that is equal to any of their dynamic domains.
 In those cases, we simply refer to their \emph{domain}.
 
-Associating a set of values to a type is done by first substituting its configuration variables
-and subprogram parameters by their values in the dynamic environment. This is done via the
-partial function $\substparams$ (see \secref{TypingRule.SubstituteParameterValues}).
+Associating a set of values to a type is done by evaluating any expressions appearing
+in the type definitions via the ASL semantics~\cite{ASLSemanticsReference},
+specifically via the relation
+\[
+  \evalexprsef{\overname{\envs}{\env} \aslsep \overname{\expr}{\ve}} \;\aslrel\;
+  \Normal(\overname{\vals}{\vv}\aslsep\overname{\XGraphs}{\vg}) \cup
+  \overname{\TError}{\ErrorConfig}
+\]
+which evaluates side-effect-free expressions and either returns
+a configuration of the form $\Normal(\vv,\vg)$ or a dynamic error.
+In the first case, $\vv$ is a native value and $\vg$
+is an \emph{execution graph}. Execution graphs are related to the concurrent semantics
+and can be ignored in the context of defining dynamic domains.
+In the latter case (which can occur if, for example, an expression attempts to divide
+\texttt{8} by \texttt{3}), a dynamic error configuration for which we use the notation
+$\ErrorConfig$ is returned.
 %
-Then, statically evaluating expressions that appear inside types is done via the
-function $\staticeval$ (see \chapref{staticevaluation}).
+The dynamic domain is undefined in cases where evaluating side-effect-free expressions
+results in a dynamic error.
+%
+The dynamic domain is also undefined if the given type $\vt$ is not well-typed in $\tenv$.
+That is, if $\annotatetype{\tenv, \vt} \typearrow \TypeErrorConfig$.
+
+As part of the definition, we also associate dynamic domains to integer constraints
+by overloading $\dynamicdomain$:
+\[
+  \dynamicdomain : \overname{\envs}{\env} \times \overname{\intconstraint}{\vc}
+  \partialto \overname{\pow{\vals}}{\vd}
+\]
+
+\subsection{Prose}
+For an environment $\env \in \envs$ and a type $\vt$, the domain is $\vd$ and one of the following holds:
```

Comment:
throughout:
one of the following holds -> one of the following applies

all of the following holds -> all of the following apply